### PR TITLE
feat(chain): create BlockHeight & UnixSeconds struct

### PR DESCRIPTION
**Pull request by 15IITian**
bitcoindevkit/bdk_wallet#147

### CODE REVIEW

**Comment by 15IITian** [comment]
Please tell any changes required till so far?

**Comment by 15IITian** [comment]
image

I created this command to directly subtract height value from a variable of type BlockHeight ->
so should I create two commands which will give output in `u32` and `BlockHeight` respectively.
as in various places sometimes we need result in either types .

**Comment by notmandatory** [comment]
This all looks good to me, be sure to fix the docs and clippy errors. Then also check where these new types should be used in the chain module code. Thanks!

**Comment by 15IITian** [comment]
> Then also check where these new types should be used in the chain module code

But IMO, I have to make crates other than `bdk-chain` to be compatible with these structs  as they are using `bdk-chain` structs in them.

File: crates/chain/src/lib.rs [hunk 1]
```diff
@@ -97,5 +100,58 @@ pub mod collections {
...
+}
+impl Default for UnixSeconds {
+    fn default() -> Self {
+        UnixSeconds(0)
+    }
+}
+
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq, Copy, PartialOrd, Ord, core::hash::Hash)]
+pub struct BlockHeight(Height);
```
Comment (evanlinjin) [inline crates/chain/src/lib.rs:121]: What is the rationale of having another type that wraps `bitcoin::absolute::Height`?

Comment (15IITian) [reply crates/chain/src/lib.rs:121]: > What is the rationale of having another type that wraps `bitcoin::absolute::Height`?

As Height struct in rust-bitcoin neither have `Default` trait implemention and  nor we can  implement it in our codebase.
Thus wrapped it .

**Comment by 15IITian** [comment]
https://github.com/bitcoindevkit/bdk/blob/d77a7f2ff13419d394e6a19f9942344a7c19a724/crates/chain/tests/test_local_chain.rs#L73

In the `test` folder of `bdk-chain` crate ->  
There is redundancy of  creating  `hash` and `height` of same values like this function has .
Suggestion: 
   * Either  in each function , we can define some varibles containing these values  and use them like this -> 
   
image

   (But still redundant at the function level)
  * Or make  const variables containing them.

File: crates/chain/src/lib.rs (hunk 1 above)
Comment (evanlinjin) [reply crates/chain/src/lib.rs:121]: Thanks for the explanation, this probably requires a doc comment.

Comment (15IITian) [reply crates/chain/src/lib.rs:121]: Raised an issue for lack of this trait implementation  in `rust-bitcoin`

https://github.com/rust-bitcoin/rust-bitcoin/issues/2532

**Comment by 15IITian** [comment]
Once this  https://github.com/rust-bitcoin/rust-bitcoin/pull/2533 for adding Default trait implementation  in rust-bitcoin will  get successfully merged.
Both `locktime::absolute::Height` and  `locktime::absolute::Time` will have their default implementation but these changes will get reflected in their future releases. 
So Should we skip the idea of using `Height` struct  for now?

**Comment by 15IITian** [comment]
image

It would be solved after `Height` & `Time` structs will have their default implementation ,which IMO would be reflected in  their future releases.
So should we stop this work for now ?

**Comment by notmandatory** [comment]
@15IITian yes it's fine to put this PR on hold until rust-bitcoin/rust-bitcoin#2533 is merged.

File: crates/chain/src/lib.rs [hunk 2]
```diff
@@ -97,5 +100,58 @@ pub mod collections {
     pub use hashbrown::hash_map;
 }
 
+/// Number of seconds that have elapsed since 00:00:00 UTC on 1 January 1970, the Unix epoch.
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq, Copy, PartialOrd, Ord, core::hash::Hash)]
+pub struct UnixSeconds(u64);
```
Comment (Kixunil) [inline crates/chain/src/lib.rs:106]: `bitcoin::absolute::Time` may be more appropriate.

File: crates/chain/src/lib.rs [hunk 3]
```diff
@@ -97,5 +100,58 @@ pub mod collections {
...
+
+impl Default for BlockHeight {
+    fn default() -> Self {
+        return BlockHeight(
+            Height::from_consensus(0).expect("Failed to create a default BlockHeight"),
+        );
+    }
+}
+
+impl Deref for BlockHeight {
```
Comment (Kixunil) [inline crates/chain/src/lib.rs:131]: [This is an anti-pattern](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html)

File: crates/chain/src/lib.rs (hunk 2 above)
Comment (15IITian) [reply crates/chain/src/lib.rs:106]: > `bitcoin::absolute::Time` may be more appropriate.


image

Comment (Kixunil) [reply crates/chain/src/lib.rs:106]: Well clearly the default is ambiguous so if there's a sensible default in your struct you should implement it manually and if there isn't you shouldn't implement it.

Comment (notmandatory) [reply crates/chain/src/lib.rs:106]: > `bitcoin::absolute::Time` may be more appropriate.

Good point, I didn't see `Time` and `Height` are both exposed under `bitcoin::absolute::`. Should certainly use those paths for `bdk_chain`.

**Comment by notmandatory** [comment]
> @15IITian yes it's fine to put this PR on hold until [rust-bitcoin/rust-bitcoin#2533](https://github.com/rust-bitcoin/rust-bitcoin/pull/2533) is merged.

You should also be able to continue this PR using `bitcoin::absolute::Time` and `bitcoin::absolute::Height` without those types having implemented default values yet. You can adapt any places that expect defaults to use `Time::MIN` and `Height::MIN`. Then if/when these types get defaults it's only a small internal change that should not affect any of our APIs.

**Comment by notmandatory** [comment]
@15IITian I created bitcoindevkit/bdk#1372 to switch the u64s for Time and I'll also replace the u32's with Height so I'm closing this PR.

**Pull request closed without merging**